x-pack/plugins/security_solution/public/timelines/components/timeline/unified_components/data_table/index.tsx

```diff
@@ -0,0 +1,360 @@
+import { timelineDefaults } from '../../../../store/defaults';
+import { timelineActions } from '../../../../store';
+import { transformTimelineItemToUnifiedRows } from '../utils';
+
+export const SAMPLE_SIZE_SETTING = 500;
+const DataGridMemoized = React.memo(UnifiedDataTable);
+
+type CommonDataTableProps = {
+  columns: ColumnHeaderOptions[];
+  rowRenderers: RowRenderer[];
+  timelineId: string;
+  itemsPerPage: number;
+  itemsPerPageOptions: number[];
+  events: TimelineItem[];
+  refetch: inputsModel.Refetch;
+  onFieldEdited: () => void;
+  totalCount: number;
+  onEventClosed: (args: ToggleDetailPanel) => void;
+  expandedDetail: ExpandedDetailTimeline;
+  showExpandedDetails: boolean;
+  onChangePage: OnChangePage;
+  activeTab: TimelineTabs;
+  dataLoadingState: DataLoadingState;
+  updatedAt: number;
+  isTextBasedQuery?: boolean;
+} & Pick<UnifiedDataTableProps, 'onSort' | 'onSetColumns' | 'sort' | 'onFilter'>;
+
+interface DataTableProps extends CommonDataTableProps {
+  dataView: DataView;
+}
+
+export const TimelineDataTableComponent: React.FC<DataTableProps> = memo(
+  function TimelineDataTableMemo({
+    columns,
+    dataView,
+    activeTab,
+    timelineId,
+    itemsPerPage,
+    itemsPerPageOptions,
+    rowRenderers,
+    sort,
+    events,
+    onFieldEdited,
+    refetch,
+    dataLoadingState,
+    totalCount,
+    onEventClosed,
+    showExpandedDetails,
+    expandedDetail,
+    onChangePage,
+    updatedAt,
+    isTextBasedQuery = false,
+    onSetColumns,
+    onSort,
+    onFilter,
+  }) {
+    const dispatch = useDispatch();
+
+    // Store context in state rather than creating object in provider value={} to prevent re-renders caused by a new object being created
+    const [activeStatefulEventContext] = useState({
+      timelineID: timelineId,
+      enableHostDetailsFlyout: true,
+      enableIpDetailsFlyout: true,
+      tabType: activeTab,
+    });
+
+    const {
+      services: {
+        uiSettings,
+        fieldFormats,
+        storage,
+        dataViewFieldEditor,
+        notifications: { toasts: toastsService },
+        theme,
+        data: dataPluginContract,
+      },
+    } = useKibana();
+
+    const [expandedDoc, setExpandedDoc] = useState<DataTableRecord & TimelineItem>();
+    const [fetchedPage, setFechedPage] = useState<number>(0);
+
+    const { browserFields, runtimeMappings } = useSourcererDataView(SourcererScopeName.timeline);
+
+    const showTimeCol = useMemo(() => !!dataView && !!dataView.timeFieldName, [dataView]);
+
+    const tableSettings = useMemo(() => {
+      const columnSettings = columns.reduce((acc, item) => {
+        if (item.initialWidth) {
+          acc[item.id] = { width: item.initialWidth };
+        }
+        return acc;
+      }, {} as Record<string, UnifiedDataTableSettingsColumn>);
+
+      return {
+        columns: columnSettings,
+      };
+    }, [columns]);
+
+    const defaultColumnIds = useMemo(() => columns.map((c) => c.id), [columns]);
```

Comment:
Yea. I wonder how fast the app will be, if we use selectors to get the column id information from redux instead and all the other transformations/filtering we have going on here

Reply:
Sorry I do not understand. In Redux also we store columns array. To get the ids, we will still have to loop through the list of columns, no? I don't think redux caches the logic of selectors. 

For the follow up PR, we can get together to see more optimization oppurtunities. What do you think?

Reply:
Given the same input, the computation done in a selector is cached. So if the id's are requested in multiple places, if the redux state storing the columns hasn't actually changed, then the loop is only ever done once.

Reply:
Okay so in follow up PR I will move it to the selectors.